[RLlib; docs] RLlib documentation do-over (new API stack): Main index page. (#48285)
ray-project/ray: doc/source/rllib/rllib-examples.rst

   that closely match (or are identical) to the previously seen observations.

- `How to train with nested observation spaces <https://github.com/ray-project/ray/blob/master/rllib/examples/connectors/nested_observation_spaces.py>`__:
   Learning in arbitrarily nested observation spaces
   (using a CartPole-v1 variant with a nested Dict observation space).

Curriculum Learning
-------------------

-  `How to set up curriculum learning with the custom callbacks API <https://github.com/ray-project/ray/blob/master/rllib/examples/curriculum/curriculum_learning.py>`__:
    Example of how to make the environment go through different levels of difficulty (from easy to harder to solve)
    and thus help the learning algorithm to cope with an otherwise unsolvable task.
    Also see the :doc:`curriculum learning how-to </rllib/rllib-advanced-api>` from the documentation.

Environments
------------
- `How to register a custom gymnasium environment <https://github.com/ray-project/ray/blob/master/rllib/examples/envs/custom_gym_env.py>`__:
   Example showing how to write your own RL environment using ``gymnasium`` and register it to run train your algorithm against this env with RLlib.

> - `How to set up rendering (and recording) of the environment trajectories during training with WandB <https://github.com/ray-project/ray/blob/master/rllib/examples/envs/env_rendering_and_recording.py>`__:

Reviewer: ```suggestion
- `How to set up rendering and recording of the environment trajectories during training with W&B <https://github.com/ray-project/ray/blob/master/rllib/examples/envs/env_rendering_and_recording.py>`__:
```

   side.

Inference (of Models/Policies)
------------------------------

- `How to do inference with an already trained policy <https://github.com/ray-project/ray/blob/master/rllib/examples/inference/policy_inference_after_training.py>`__:
   Example of how to perform inference (compute actions) on an already trained policy.
- `How to do inference with an already trained (LSTM) policy <https://github.com/ray-project/ray/blob/master/rllib/examples/inference/policy_inference_after_training_with_lstm.py>`__:
   Example of how to perform inference (compute actions) on an already trained (LSTM) policy.

Multi-Agent RL
--------------

- `How to set up independent multi-agent training <https://github.com/ray-project/ray/blob/master/rllib/examples/multi_agent/pettingzoo_independent_learning.py>`__:
   Set up RLlib to run any algorithm in (independent) multi-agent mode against a multi-agent environment.
- `How to set up shared-parameter multi-agent training <https://github.com/ray-project/ray/blob/master/rllib/examples/multi_agent/pettingzoo_parameter_sharing.py>`__:
   Set up RLlib to run any algorithm in (shared-parameter) multi-agent mode against a multi-agent environment.

> - `How to compare a heuristic policy vs a trained one on rock-paper-scissors <https://github.com/ray-project/ray/blob/master/rllib/examples/multi_agent/rock_paper_scissors_heuristic_vs_learned.py>`__ and `Rock-paper-scissors learned vs learned <https://github.com/ray-project/ray/blob/master/rllib/examples/multi_agent/rock_paper_scissors_learned_vs_learned.py>`__:

Reviewer: ```suggestion
- `How to compare a heuristic policy versus a trained one on rock-paper-scissors <https://github.com/ray-project/ray/blob/master/rllib/examples/multi_agent/rock_paper_scissors_heuristic_vs_learned.py>`__ and `Rock-paper-scissors learned vs learned <https://github.com/ray-project/ray/blob/master/rllib/examples/multi_agent/rock_paper_scissors_learned_vs_learned.py>`__:
```

   Two examples of different heuristic and learned policies competing against each other in the rock-paper-scissors environment.
- `How to use agent grouping in a multi-agent environment (two-step game) <https://github.com/ray-project/ray/blob/master/rllib/examples/multi_agent/two_step_game_with_grouped_agents.py>`__:
   Example on how to use agent grouping in a multi-agent environment (the two-step game from the `QMIX paper <https://arxiv.org/pdf/1803.11485.pdf>`__).

> - `How to set up multi-agent training vs a PettingZoo environment <https://github.com/Farama-Foundation/PettingZoo/blob/master/tutorials/Ray/rllib_pistonball.py>`__:

Reviewer: ```suggestion
- `How to set up multi-agent training versus a PettingZoo environment <https://github.com/Farama-Foundation/PettingZoo/blob/master/tutorials/Ray/rllib_pistonball.py>`__:
```

   Example on how to use RLlib to learn in `PettingZoo <https://www.pettingzoo.ml>`__ multi-agent environments.
- `How to hand-code a (heuristic) policy <https://github.com/ray-project/ray/blob/master/rllib/examples/multi_agent/custom_heuristic_policy.py>`__:
   Example of running a custom hand-coded policy alongside trainable policies.

> - `How to train a single policy (weight sharing) controlling more than one agents <https://github.com/ray-project/ray/blob/master/rllib/examples/multi_agent/multi_agent_cartpole.py>`__:

Reviewer: ```suggestion
- `How to train a single policy (weight sharing) controlling more than one agent <https://github.com/ray-project/ray/blob/master/rllib/examples/multi_agent/multi_agent_cartpole.py>`__:
```

   Example of how to define weight-sharing layers between two different policies.

Ray Serve and RLlib
-------------------

- `How to use a trained RLlib algorithm with Ray Serve <https://github.com/ray-project/ray/tree/master/rllib/examples/ray_serve/ray_serve_with_rllib.py>`__
   This script offers a simple workflow for 1) training a policy with RLlib first, 2) creating a new policy 3) restoring its weights from the trained
   one and serving the new policy with Ray Serve.

Ray Tune and RLlib
------------------
- `How to define a custom progress reporter and use it with Ray Tune and RLlib <https://github.com/ray-project/ray/blob/master/rllib/examples/ray_tune/custom_progress_reported.py>`__:
   Example of how to write your own progress reporter (for a multi-agent experiment) and use it with Ray Tune and RLlib.

- `How to define and plug in your custom logger into Ray Tune and RLlib <https://github.com/ray-project/ray/blob/master/rllib/examples/ray_tune/custom_logger.py>`__:
   How to setup a custom Logger object in RLlib and use it with Ray Tune.

> - `How to Custom tune experiment <https://github.com/ray-project/ray/blob/master/rllib/examples/ray_tune/custom_experiment.py>`__:

Reviewer: ```suggestion
- `How to custom tune an experiment <https://github.com/ray-project/ray/blob/master/rllib/examples/ray_tune/custom_experiment.py>`__:
```
